scripts/image_builder.py

```diff
@@ -0,0 +1,797 @@
+    cmd: str
+
+    def emit(self, file: TextIO, enc: PloScriptEncoding, payload_offs: int, is_relative: bool) -> Tuple[int, Optional[ProgInfo]]:
+        if enc == PloScriptEncoding.DEBUG_ASDICT:
+            file.write(str(asdict(self)) + "\n")
+        elif enc == PloScriptEncoding.STRING_MAGIC_V1:
+            # basic plo cmd - just emit as a string
+            file.write(self.cmd + "\n")
+        else:
+            raise NotImplementedError(f"PloScriptEncoding {enc.value} not implemented")
+
+        return payload_offs, None
+
+
+@dataclass
+class PloCmdAlias(PloCmdBase):
+    """Alias command - used for memory aliases - emitted usually by app/kernel/blob/call commands"""
+    NAME: ClassVar = "alias"
+    name: str = field(default=NAME, kw_only=True)
+
+    filename: str               # target alias filename
+    size: int                   # target alias size
+    set_base: bool = False      # should we set new base before aliasing?
+
+    def emit(self, file: TextIO, enc: PloScriptEncoding, payload_offs: int, is_relative: bool) -> Tuple[int, Optional[ProgInfo]]:
+        if enc == PloScriptEncoding.DEBUG_ASDICT:
+            file.write(str(asdict(self)) + "\n")
+        elif enc == PloScriptEncoding.STRING_MAGIC_V1:
+            if self.set_base:
+                if is_relative:
+                    file.write(f"alias -rb {payload_offs:#x}\n")
+
+                    # TODO: add real support for virtual base change (payload_offs in partition, virtual_offs as relative alias)
+                    # for now we'll just reset payload_offs - any relative `app` after base change in the same script would break
+                    payload_offs = 0
+                else:
+                    file.write(f"alias -b {payload_offs:#x}\n")
+
+            aliasCmd = "alias -r" if is_relative else "alias"
+            file.write(f"{aliasCmd} {self.filename} {payload_offs:#x} {self.size:#x}\n")
+        else:
+            raise NotImplementedError(f"PloScriptEncoding {enc.value} not implemented")
+
+        return payload_offs + round_up(self.size, SIZE_PAGE), ProgInfo(PREFIX_PROG_STRIPPED / self.filename, payload_offs, self.size)
+
+
+@dataclass
+class PloCmdKernel(PloCmdBase):
+    """Kernel is a special type of 'app' due to kernelimg command subtype:
+        kernel[img] <device>
+        kernel flash0
+    """
+    NAME: ClassVar = "kernel"
+    device: str
+
+    # internal fields
+    suffix: str = field(default="elf", init=False)
+    size: int = field(init=False)
+    filename: str = field(init=False)
+    abspath: Path = field(init=False)
+
+    def __post_init__(self, extra_flags: str = ''):
+        if self.name == "kernelimg":
+            self.suffix = "bin"
+
+        self.filename = f"phoenix-{'-'.join(TARGET.split('-')[:2])}.{self.suffix}"
+        self.abspath = PREFIX_PROG_STRIPPED / self.filename
+        self.size = os.path.getsize(self.abspath)
```

Comment:
I'm not entirely convinced `self.abspath.stat().st_size` is more semantic, but I'm new to `pathlib`. Probably I should also use `path.open()` instead of `open(path)`?

Reply:
I'll keep it as-is for now, style changes can be introduced later